application/src/ext/java/org/opentripplanner/ext/carpooling/filter/DirectionalCompatibilityFilter.java

```diff
@@ -0,0 +1,120 @@
+   * Default maximum bearing difference for compatibility.
+   * 60° allows for reasonable detours while preventing perpendicular or opposite directions.
+   */
+  public static final double DEFAULT_BEARING_TOLERANCE_DEGREES = 60.0;
+
+  private final double bearingToleranceDegrees;
+
+  public DirectionalCompatibilityFilter() {
+    this(DEFAULT_BEARING_TOLERANCE_DEGREES);
+  }
+
+  public DirectionalCompatibilityFilter(double bearingToleranceDegrees) {
+    this.bearingToleranceDegrees = bearingToleranceDegrees;
+  }
+
+  @Override
+  public boolean accepts(
+    CarpoolTrip trip,
+    WgsCoordinate passengerPickup,
+    WgsCoordinate passengerDropoff
+  ) {
+    List<WgsCoordinate> routePoints = trip.routePoints();
+
+    if (routePoints.size() < 2) {
+      LOG.warn("Trip {} has fewer than 2 route points, rejecting", trip.getId());
+      return false;
+    }
+
+    double passengerBearing = DirectionalCalculator.calculateBearing(
+      passengerPickup,
+      passengerDropoff
+    );
+
+    for (int i = 0; i < routePoints.size() - 1; i++) {
+      if (isSegmentCompatible(routePoints.get(i), routePoints.get(i + 1), passengerBearing)) {
+        LOG.debug(
+          "Trip {} accepted: passenger journey aligns with segment {} ({} to {})",
+          trip.getId(),
+          i,
+          routePoints.get(i),
+          routePoints.get(i + 1)
+        );
+        return true;
+      }
+    }
+
+    // Check full route as fallback (handles complex multi-segment compatibility)
+    if (
+      isSegmentCompatible(
+        routePoints.get(0),
+        routePoints.get(routePoints.size() - 1),
+        passengerBearing
+      )
+    ) {
+      LOG.debug(
+        "Trip {} accepted: passenger journey aligns with full route ({} to {})",
+        trip.getId(),
+        routePoints.get(0),
+        routePoints.get(routePoints.size() - 1)
+      );
+      return true;
+    }
+
+    LOG.debug(
+      "Trip {} rejected by directional filter: passenger journey (bearing {}°) not aligned with any route segments",
+      trip.getId(),
+      Math.round(passengerBearing)
+    );
+    return false;
+  }
+
+  /**
+   * Checks if a segment is directionally compatible with the passenger journey.
+   *
+   * @param segmentStart Start coordinate of the segment
+   * @param segmentEnd End coordinate of the segment
+   * @param passengerBearing Bearing of passenger journey
+   * @return true if segment bearing is within tolerance of passenger bearing
+   */
+  private boolean isSegmentCompatible(
+    WgsCoordinate segmentStart,
+    WgsCoordinate segmentEnd,
+    double passengerBearing
+  ) {
+    double segmentBearing = DirectionalCalculator.calculateBearing(segmentStart, segmentEnd);
+
+    double bearingDiff = DirectionalCalculator.bearingDifference(segmentBearing, passengerBearing);
+
+    return bearingDiff <= bearingToleranceDegrees;
+  }
+
+  /**
+   * Gets the configured bearing tolerance in degrees.
+   */
+  public double getBearingToleranceDegrees() {
```

Comment:
This can be package local and in general we try to sort our methods according to https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/doc/dev/decisionrecords/Codestyle.md#sorting-class-members